### DISCUSSION TITLE
Blocking Queues in F'

### DISCUSSION THREAD

**Discussion by Gonta01** [nasa/fprime#2920] [Q&A] [answered]
This issue is related to [FreeRTOS eventGroups, semaphores in F'](https://github.com/nasa/fprime/discussions/2851)
@LeStarch suggested to use Os::Mutex as a substitution for eventGroups, semaphores but I used messageQueue instead.
Here is example:

Inside active component is used send_data():

void send_data() {
	if (startSend()) {
		waitSendEnd();
	}
}

startSend initiates hardware data sending;

waitSendEnd waits until data sending ends.

void waitSendEnd {
	queue.receive(); // waits in blocking mode since there is no way in OSAL queue to set timeout whereas FreeRTOS gives such opportunity
}

SendEndHandler {
	queue.send();
}

Here is problem: if sendEndHandler didn't work (interrupt didn't finished for some hardware reason), the system will collapse because of invocation overflow (rateGroup will invocate run handler but the task is blocked because of waitSendEnd())

Similarly for receive_data:
void receive_data() {
	if (startReception()) {
		waitReceptionEnd();
	}
}

If nothing will be received the whole system will be collapsed because task is suspended and wait for reception of some data.

The questions are:
1) Why doesn't F' OSAL give ability to use timeout for queue receive?
2) Does approach above is compatible with F' ideology? What are alternatives?
3) Could Os::Mutex be used here to resolve this as it was suggested in another issue. Originally instead of queue was used eventGroups.

**Comment by LeStarch** [comment, marked as answer]
1.  F´ queues where built to be the basis of an event loop of the form:
```
while (message != EXIT) {
    m = q.receive
    process(m)
}
// End of FSW
```

For this use-case no timeout is needed as it will just try to receive again.

2. @timcanham can you answer this?
3. Theoretically you could lock a mutex while sending the data.  This would ensure that the other receive_data handler is in one of three states:
  1. Not started
  2. Data sent
  3. Data sent, but erred
You could set a state variable and track in `receive_data`

**Reply by timcanham** [reply]
2. I don't think timeouts are against F Prime ideology, but the primary use-case is within the message loop as @LeStarch says. I could see it as an optional parameter for the `receive()` call, where F Prime components could use the block-forever option (usually timeout of zero). If we have unknown timeouts, it is hazardous to stall the message handling loop as @Gonta01 has stated. There are two other options as I see it. Send the message, and have the destination do a callback through a port. The timeout can be done by: 1) A scheduler call if the scale is on the order of seconds. 2) A timer callback. Both of these would free up the message queue.

**Reply by Gonta01** [reply]
@timcanham "I could see it as an optional parameter for the receive() call, where F Prime components could use the block-forever option (usually timeout of zero)." Could you please give a hint how to implement it. I'm trying to substitute fprime/Os/Queue.hpp with my own that is placed in my folder but without success now.

**Reply by timcanham** [reply]
I think this feature would have to propagate to the delegate, and the delegate would have to provide (or disallow) the feature based on the API for the given OS or implementation.

**Reply by Gonta01** [reply]
@timcanham thank you for your response but to be honest I didn't understand what do you mean. Are there any example of such approach?

**Reply by timcanham** [reply]
I'm rereading the thread now... I originally thought you meant have a timeout on the send() side of a message queue. That's a possible pattern, where you specify that the send() blocks but with a timeout. For receive(), it's a different matter. In your case, you want to use a message queue apart from the usual F Prime component queue for hardware sequencing. I see how you would want to use a queue with a timeout. This is how I would do it instead:

1) Based on your overflow worries, I'm guessing that your timeout is some multiple of the rate group invocation. For example, let's say your rate group is 1Hz, and your timeout is 5 seconds.
2) You can make the queue non-blocking. 
3) Start the hardware sequence (your startSend())
4) Instead of blocking on the receive(), make a call to [`getMessagesAvailable()`](https://github.com/nasa/fprime/blob/devel/Os/Queue.hpp#L118) in your run handler. If it is non-zero, you know you have a message waiting, and you can call [`receive()`](https://github.com/nasa/fprime/blob/devel/Os/Queue.hpp#L107). If it is zero, decrement the remaining timeout by 1 second.
5) If it times out, you can detect it when the timeout count reaches zero.
